### DISCUSSION TITLE
Extracting Steam TOTP from the Steam 3.0 Android app

### DISCUSSION THREAD

**Discussion by null-dev** [keepassxreboot/keepassxc#9591] [Ideas]
I don't see this documented yet so I'm going to document it here.

Here's how you extract the Steam TOTP secret from the Steam 3.0 Android app. The easiest way is to just use a rooted phone but there are other methods (not mentioned here because they are quite complicated).

1. Follow the instructions here: https://github.com/JustArchiNET/ArchiSteamFarm/discussions/2786
2. You will end up with something like this:

  ```
  {
    "accounts": {
      "76561197960287930": {
        "shared_secret": "9jMbHnerWZPtmzjdyEexyEnyHS8=",
        "identity_secret": "****************************",
        "secret_1": "****************************",
        "serial_number": "********************",
        "revocation_code": "******",
        "account_name": "Rabscuttle",
        "token_gid": "****************",
        "steamguard_scheme": 2,
        "steamid": "76561197960287930"
      }
    }
  }
  ```
3. The field that you are interested in is the `shared_secret` field, this is your TOTP secret. You can't use it directly however, as it's in Base64. You need to convert it into Base32.
    - On Linux/Mac, you can use the following command: `printf 'PUT YOUR shared_secret HERE' | base64 -d | base32`
      - Example usage: `printf '9jMbHnerWZPtmzjdyEexyEnyHS8=' | base64 -d | base32`
    - On Windows you can use an online tool (although for security, I wouldn't recommend it). [Here's a sample tool.](https://ciphereditor.com/share#blueprint=eyJ0eXBlIjoiYmx1ZXByaW50IiwicHJvZ3JhbSI6eyJ0eXBlIjoicHJvZ3JhbSIsIm9mZnNldCI6eyJ4IjoxNjgsInkiOi0zNX0sImZyYW1lIjp7IngiOi0xNjAsInkiOi05Niwid2lkdGgiOjMyMCwiaGVpZ2h0IjoxOTJ9LCJjaGlsZHJlbiI6W3sidHlwZSI6ImNvbnRyb2wiLCJpZCI6IjIxIiwibGFiZWwiOiJDb250cm9sIiwidmFsdWUiOiI5ak1iSG5lcldaUHRtempkeUVleHlFbnlIUzg9IiwidmlzaWJpbGl0eSI6ImV4cGFuZGVkIiwiZnJhbWUiOnsieCI6LTE4NSwieSI6LTM4NCwid2lkdGgiOjMyMCwiaGVpZ2h0IjoxMjZ9fSx7InR5cGUiOiJvcGVyYXRpb24iLCJuYW1lIjoiQGNpcGhlcmVkaXRvci9leHRlbnNpb24tZXNzZW50aWFscy9iaW5hcnktdG8tdGV4dCIsImV4dGVuc2lvblVybCI6Imh0dHBzOi8vY2RuLmNpcGhlcmVkaXRvci5jb20vZXh0ZW5zaW9ucy9AY2lwaGVyZWRpdG9yL2V4dGVuc2lvbi1lc3NlbnRpYWxzLzEuMC4wLWFscGhhLjEvZXh0ZW5zaW9uLmpzIiwicHJpb3JpdHlDb250cm9sTmFtZXMiOlsiZW5jb2RlZERhdGEiLCJkYXRhIiwiYWxwaGFiZXQiLCJwYWRkaW5nIl0sImZyYW1lIjp7IngiOi0xOTEsInkiOi0xOTMsIndpZHRoIjozMjAsImhlaWdodCI6MjM4fSwiY29udHJvbHMiOnsiZGF0YSI6eyJpZCI6IjI0IiwidmFsdWUiOnsidHlwZSI6ImJ5dGVzIiwiZGF0YSI6IjlqTWJIbmVyV1pQdG16amR5RWV4eUVueUhTOEEifX0sImVuY29kZWREYXRhIjp7ImlkIjoiMjciLCJ2YWx1ZSI6IjlqTWJIbmVyV1pQdG16amR5RWV4eUVueUhTOD0ifX19LHsidHlwZSI6Im9wZXJhdGlvbiIsIm5hbWUiOiJAY2lwaGVyZWRpdG9yL2V4dGVuc2lvbi1lc3NlbnRpYWxzL2JpbmFyeS10by10ZXh0IiwiZXh0ZW5zaW9uVXJsIjoiaHR0cHM6Ly9jZG4uY2lwaGVyZWRpdG9yLmNvbS9leHRlbnNpb25zL0BjaXBoZXJlZGl0b3IvZXh0ZW5zaW9uLWVzc2VudGlhbHMvMS4wLjAtYWxwaGEuMS9leHRlbnNpb24uanMiLCJwcmlvcml0eUNvbnRyb2xOYW1lcyI6WyJkYXRhIiwicGFkZGluZyIsImFscGhhYmV0IiwiZW5jb2RlZERhdGEiXSwiZnJhbWUiOnsieCI6LTE1MSwieSI6NTksIndpZHRoIjozMjAsImhlaWdodCI6Mjg2fSwiY29udHJvbHMiOnsiZGF0YSI6eyJpZCI6IjMwIiwidmFsdWUiOnsidHlwZSI6ImJ5dGVzIiwiZGF0YSI6IjlqTWJIbmVyV1pQdG16amR5RWV4eUVueUhTOEEifX0sImFscGhhYmV0Ijp7InZhbHVlIjoiQUJDREVGR0hJSktMTU5PUFFSU1RVVldYWVoyMzQ1NjciLCJ2aXNpYmlsaXR5IjoiZXhwYW5kZWQifSwicGFkZGluZyI6eyJ2YWx1ZSI6IiJ9LCJlbmNvZGVkRGF0YSI6eyJpZCI6IjMzIiwidmFsdWUiOiI2WVpSV0hUWFZOTVpIM00zSERPNFFSNVJaQkU3RUhKUEFBIn19fSx7InR5cGUiOiJjb250cm9sIiwiaWQiOiIzNSIsImxhYmVsIjoiQ29udHJvbCIsInZhbHVlIjoiNllaUldIVFhWTk1aSDNNM0hETzRRUjVSWkJFN0VISlBBQSIsInZpc2liaWxpdHkiOiJleHBhbmRlZCIsImZyYW1lIjp7IngiOjE5MywieSI6LTM4MSwid2lkdGgiOjMyMCwiaGVpZ2h0IjoxNTB9fSx7InR5cGUiOiJ2YXJpYWJsZSIsImF0dGFjaG1lbnRzIjpbIjIxIiwiMjciXX0seyJ0eXBlIjoidmFyaWFibGUiLCJhdHRhY2htZW50cyI6WyIyNCIsIjMwIl19LHsidHlwZSI6InZhcmlhYmxlIiwiYXR0YWNobWVudHMiOlsiMzMiLCIzNSJdfV19fQ) You put your `shared_secret` in the top left "Control" that says "Push". The output in the top right "Control" that says "Pull" will be your TOTP secret.

**Comment by Barmem** [comment]
What are other methods? Don't wanna actually root my phone and wanna keep a copy of authenticator on my phone.

**Reply by null-dev** [reply]
If you don't have root, you can follow the same guide linked except instead of installing frida on the phone, you [add frida to the Steam app APK](https://koz.io/using-frida-on-android-without-root/).

There is a big caveat however: Since the Steam app has been modified, you can't update it from the play store anymore. You have to generate a new modified version of the Steam app every time you want to update.

**Comment by wyfang** [comment]
mac tip: 
zsh: command not found: base32

**Comment by xperia-droid** [comment]
For Windows Users:

Save this code to an .cmd file on your PC:
```
@echo off
Rem Make powershell read this file, skip a number of lines, and execute it.
Rem This works around .ps1 bad file association as non executables.
PowerShell -Command  "$s = GC '%~dpnx0' | Select -Skip 5 | Out-String; $s='cd ""%~pd0"";'+$s; IEX $s"
goto :eof
# Start of PowerShell script here
$csharp = @"
public static class b32
{
    public static string BytesToBase32(byte[] bytes) {
        const string alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
        string output = "";
        for (int bitIndex = 0; bitIndex < bytes.Length * 8; bitIndex += 5) {
            int dualbyte = bytes[bitIndex / 8] << 8;
            if (bitIndex / 8 + 1 < bytes.Length)
                dualbyte |= bytes[bitIndex / 8 + 1];
            dualbyte = 0x1f & (dualbyte >> (16 - bitIndex % 8 - 5));
            output += alphabet[dualbyte];
        }
        return output;
    }
}
![Steamguardtoken](https://github.com/keepassxreboot/keepassxc/assets/24463981/eb906852-d14a-4bf9-a081-0a77c74bd8c1)

"@
Add-Type -TypeDefinition $csharp
$encodedSecret = Read-Host -Prompt ("Steam Guard Base64 Token") 
$decodedSecret = [System.Convert]::FromBase64String($encodedSecret)
$b32 = [b32]::BytesToBase32($decodedSecret)
Write-Host ("Steam Guard Base32 Token Result: " + $b32)
pause
```

start the .cmd file, then enter your shared_secret press enter and your token in base 32 returns:
![Steamguardtoken](https://github.com/keepassxreboot/keepassxc/assets/24463981/74a971f3-c628-4f5f-8ae7-b3932cbf9b8f)

**Comment by xperia-droid** [comment]
If you have root and LSPosed installed on your Android phone, you can get the Steam Guard Token with the following LSPosed module:
https://github.com/YifePlayte/SteamGuardDump/releases

After installing the app, you must activate the module in LSPosed and check that the Steam app is checked. Then restart the cell phone. After restarting, open the Steam app and the Steam Guard data will automatically be copied to the phone's clipboard.

The Steam Guard data is available in a json formatted string. 
The string contains the word otpauth and you only need the part after secret= up to the & character. Copy the part between secret= and the & character into the TOTP settings of your Steam Keepass entry and you then you can use the Steamguard Token in KeepassXC :).

**Comment by lipatique** [comment]
[This](https://www.reddit.com/r/Bitwarden/comments/1auercm/updated_feb_2024_guide_extracting_steam_guard/) method still working. You can have the steam guard authenticator on your unrooted Android device as normal and still get TOTP codes to key manager. Did it few days ago on my Android 13 device.
